### PR TITLE
Move decode warmup from vLLM to metal side

### PR DESCRIPTION
**Pull request by nostojicTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/vllm/issues/311)

### Problem description
Currently the parameters for decode warmup are being created at vLLM side. Because of this, we can't run metal demo and e2e tests, so we need to move the parameter creation inside metal.

vllm changes - https://github.com/tenstorrent/vllm/pull/316

#### Model tests

vllm nightly - https://github.com/tenstorrent/tt-metal/actions/runs/22135023070
t3k demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22135035672
single card demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22135047665
galaxy demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22135921831

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -156,12 +140,20 @@ def warmup_model_prefill(
         self,
         kv_cache,
         enable_trace,
-        sampling_params=[None],
+        sample_on_device_mode=None,
+        non_greedy_decoding_on_device=False,
+        max_batch_size=1,
```
Comment (mtairum) [inline models/tt_transformers/tt/generator.py]: Is this hardcoded to 1 because TTT does not support batched prefill? Or because batched prefill is similar to doing a longer seqlen prefill (with batch 1)?

**Review by mtairum — APPROVED** [review]

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py]: Currently, there is no batched prefill in TTT, but when that gets added, the warmup method should be updated accordingly. In order to catch if new codepaths are not covered by warmup, I have created an issue related to that where I suggested ways we can catch codepaths not covered by warmup - https://github.com/tenstorrent/tt-metal/issues/37490

**Review by rdraskicTT — APPROVED** [review]

**Review by gwangTT — APPROVED** [review]
Code looks good to me.

Please run T3K demo tests in CI to double-check

**Review by viktorpusTT — APPROVED** [review]

**Review by dpopovTT — APPROVED** [review]

**Review by ssanjayTT — APPROVED** [review]

File: models/common/sampling/sampling_params.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class SamplingParams:
+    """
+    Used in Generator decode forward functions for greedy decoding / sampling on device.
+    The same data class exists in vLLM at vllm/worker/tt_model_runner.py.
```
Comment (bgoelTT) [inline models/common/sampling/sampling_params.py]: Could we then refactor the vLLM PR to import this dataclass?

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (bgoelTT) [reply models/tt_transformers/tt/generator.py]: FYI @ssinghalTT 

Batched prefill is coming to TTT

Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py]: I have created an issue to add support for batched prefill in TTT. I just saw that the `max_batch_size` parameter is unnecessary for the prefill warmup. Here is the link to the issue, will do it right after merging this PR.
https://github.com/tenstorrent/tt-metal/issues/37732

**Review by pprajapatiTT — APPROVED** [review]
Approved to unlock. But discussed with Nikola to make follow up PR to mimic demo as vllm for warm up calls.

File: models/common/warmup/__init__.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
```
Comment (skhorasganiTT) [inline models/common/warmup/__init__.py]: 2026, likewise all other new files

File: models/common/warmup/__init__.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+
+# SPDX-License-Identifier: Apache-2.0
+
+from .decode_warmup import DecodeWarmupMixin
+
+__all__ = ["DecodeWarmupMixin"]
```
Comment (skhorasganiTT) [inline models/common/warmup/__init__.py]: Please update the codeowners file with the owners of `models/common/warmup`. Owners should include yourself, me, and some or all of the models/common/sampling owners (Stuti, Djordje, Radoica, Tomasz)

File: models/common/warmup/warmup_utils.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        return sampling_configs
+
+    def _create_decode_warmup_inputs(self, max_batch_size, num_gpu_blocks):
+        tokens = torch.zeros(max_batch_size, 1, dtype=torch.int32)
+        start_pos = torch.zeros(max_batch_size, dtype=torch.int32)
+        page_table = torch.zeros(max_batch_size, num_gpu_blocks, dtype=torch.int32)
+        return tokens, start_pos, page_table
+
+    def warmup_model_decode(
```
Comment (skhorasganiTT) [inline models/common/warmup/warmup_utils.py:81]: Please add in docstring that this function is called by vLLM so we are careful about the API

File: models/common/warmup/decode_warmup.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+
+class DecodeWarmupMixin:
+    """
+    Mixin class that provides decode warmup functionality for generator classes.
+
+    This class should be inherited by any generator class that needs to warm up
+    the decode forward pass. It requires the following to be defined in the
+    inheriting class:
+    - self.decode_forward_text(): method to perform decode forward pass
```
Comment (skhorasganiTT) [inline models/common/warmup/decode_warmup.py]: decode_forward , not decode_forward_text

File: models/common/warmup/decode_warmup.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+import torch
+from loguru import logger
+
+from models.common.sampling.sampling_params import SamplingParams
+
+
+class DecodeWarmupMixin:
+    """
+    Mixin class that provides decode warmup functionality for generator classes.
```
Comment (skhorasganiTT) [inline models/common/warmup/decode_warmup.py]: Please mention used by vLLM

File: models/common/warmup/decode_warmup.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    This class should be inherited by any generator class that needs to warm up
+    the decode forward pass. It requires the following to be defined in the
+    inheriting class:
+    - self.decode_forward_text(): method to perform decode forward pass
+    """
+
+    def _create_sampling_params(self, sample_on_device_mode, non_greedy_decoding_on_device, max_batch_size, mode):
```
Comment (skhorasganiTT) [inline models/common/warmup/decode_warmup.py]: in docstring please add what non_greedy_decoding_on_device means

File: models/common/warmup/warmup_utils.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                if penalties:
+                    presence_penalty = [1.2] * max_batch_size if mode == "decode" else 1.2
+                    frequency_penalty = [1.2] * max_batch_size if mode == "decode" else 1.2
+                    repetition_penalty = [1.5] * max_batch_size if mode == "decode" else 1.5
+
+                enable_log_probs = [log_probs] * max_batch_size if mode == "decode" else log_probs
+
+                temperature = [1.0] * max_batch_size if mode == "decode" else 1.0
+                top_k = [10] * max_batch_size if mode == "decode" else 10
+                top_p = [0.9] * max_batch_size if mode == "decode" else 0.9
```
Comment (skhorasganiTT) [inline models/common/warmup/warmup_utils.py:49]: Why is this different from the current version in vLLM? It's setting list for decode but int for prefill

File: models/common/warmup/decode_warmup.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+from itertools import product
+
+import torch
+from loguru import logger
+
+from models.common.sampling.sampling_params import SamplingParams
+
+
+class DecodeWarmupMixin:
```
Comment (skhorasganiTT) [inline models/common/warmup/decode_warmup.py]: The _create_sampling_params function is used for prefill so this file and class should be renamed. Maybe warmup_utils.py and WarmupForwardMixin

File: models/demos/deepseek_v3/tt/generator.py [hunk 11]
```diff
@@ -1057,6 +1052,17 @@ def set_kv_cache(self, kv_cache_list: list[ttnn.Tensor]) -> None:
...
+    def warmup_model_decode(
+        self,
+        kv_cache,
+        enable_trace,
+        max_batch_size,
+        num_gpu_blocks,
+        sample_on_device_mode=None,
+        non_greedy_decoding_on_device=False,
+    ):
+        logger.warning("decode_forward_text is not implemented for DeepseekGenerator")
```
Comment (skhorasganiTT) [inline models/demos/deepseek_v3/tt/generator.py]: This should be deleted so the function from DecodeWarmupMixin is used. DeepSeek was already working with the decode warmup in vllm

File: models/demos/qwen25_vl/tt/generator.py [hunk 12]
```diff
@@ -208,6 +209,24 @@ def __prefill_forward_single_user_text(self, tokens, page_table, user_id, last_t
...
+        non_greedy_decoding_on_device=False,
+    ):
+        return super().warmup_model_decode(
+            kv_cache=kv_cache,
+            enable_trace=enable_trace,
+            max_batch_size=max_batch_size,
+            num_gpu_blocks=num_gpu_blocks,
+            sample_on_device_mode=sample_on_device_mode,
+            non_greedy_decoding_on_device=non_greedy_decoding_on_device,
+        )
```
Comment (skhorasganiTT) [inline models/demos/qwen25_vl/tt/generator.py]: delete. should use function from DecodeWarmupMixin

File: models/demos/qwen3_vl/tt/generator.py [hunk 13]
```diff
@@ -243,6 +244,30 @@ def __prefill_forward_single_user_text(
...
+        non_greedy_decoding_on_device=False,
+    ):
+        return super().warmup_model_decode(
+            kv_cache=kv_cache,
+            enable_trace=enable_trace,
+            max_batch_size=max_batch_size,
+            num_gpu_blocks=num_gpu_blocks,
+            sample_on_device_mode=sample_on_device_mode,
+            non_greedy_decoding_on_device=non_greedy_decoding_on_device,
+        )
```
Comment (skhorasganiTT) [inline models/demos/qwen3_vl/tt/generator.py]: delete. should use function from mixin class

File: models/tt_transformers/tt/generator.py [hunk 14]
```diff
@@ -157,12 +141,21 @@ def warmup_model_prefill(
...
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: args should match those expected by vlllm and defaults should be removed as well i think

File: models/demos/llama3_70b_galaxy/tt/generator.py
Comment (skhorasganiTT) [inline models/demos/llama3_70b_galaxy/tt/generator.py]: this function needs to call _create_sampling_params, like TTT, to match current vLLM behavior

File: models/common/warmup/warmup_utils.py (hunk 9 above)
Comment (nostojicTT) [reply models/common/warmup/warmup_utils.py:49]: The reason why decode has a list is because decode operates on batch=32 per model instance, which means we should provide sampling params as a list of 32 * data_parallel integers. The sampling params then get chunked into data_parallel number of lists of 32 integer that represent sampling params. For example : DP=4 on Llama8b -> we would create SamplingParams -> temerature = [1.0] * 32 * 4 -> in decode_forward that gets chunked into [ [1.0] * 32, [1.0] * 32, [1.0] * 32, [1.0] * 32]. On the other side, prefill works with both batch=1 and batch=32 (batched_prefill). Currently, this PR covers only batch=1, but right after merging this PR, I will refactor this function in order to support also batch=32 for prefill. Issue link: https://github.com/tenstorrent/tt-metal/issues/37732

File: models/demos/deepseek_v3/tt/generator.py (hunk 11 above)
Comment (nostojicTT) [reply models/demos/deepseek_v3/tt/generator.py]: Deepseek certainly did work with decode war,up in vLLM, but what this PR should also do is add support so demo and e2e tests in tt-metal can call the `warmup_model_decode` function and warmup the model correctly. Everything is fine for other generator classes, but for deepseek, there are some minor changes needed in order to make that compatible, and I would like to fix that in a separate PR as discussed with @pprajapatiTT. Here is the link to the issue - https://github.com/tenstorrent/tt-metal/issues/37771

File: models/demos/llama3_70b_galaxy/tt/generator.py
Comment (nostojicTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: I would like to cover that in a separate PR - https://github.com/tenstorrent/tt-metal/issues/37732

File: .github/CODEOWNERS [hunk 15]
```diff
@@ -293,6 +293,7 @@ tests/ttnn/distributed/ @tenstorrent/metalium-developers-ttnn-core @tenstorrent/
 /**/functional_*/ @uaydonat @esmalTT @tenstorrent/codeowner-bypass
 models/common @yieldthought @mtairum @uaydonat @gwangTT @tenstorrent/codeowner-bypass
 models/common/sampling/ @sraizada-tt @djordje-tt @rdraskicTT @tchedaTT @tenstorrent/codeowner-bypass
+models/common/warmup/ @nostojicTT @skhorasganiTT @sraizada-tt @djordje-tt @rdraskicTT @tchedaTT
```
Comment (skhorasganiTT) [inline .github/CODEOWNERS]: also `@tenstorrent/codeowner-bypass`

File: models/common/sampling/sampling_params.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
```
Comment (skhorasganiTT) [inline models/common/sampling/sampling_params.py]: "vllm/v1/worker/tt_model_runner.py" (vllm/worker/tt_model_runner.py will be deleted soon)

File: models/common/warmup/warmup_utils.py (hunk 9 above)
Comment (skhorasganiTT) [reply models/common/warmup/warmup_utils.py:49]: Yes what I meant was that prefill should also be list like the current behaviour. Ok if will be addressed right after

File: models/tt_transformers/tt/generator.py [hunk 17]
```diff
@@ -330,7 +319,7 @@ def prefill_forward_text(
 
         # we need this here becuase of tt-metal tests
         if warmup_prefill:
-            self.warmup_model_prefill(kv_cache, enable_trace)
+            self.warmup_model_prefill(kv_cache, enable_trace, False, False)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Should be True, True?

File: models/tt_transformers/tt/generator.py
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: This looks wrong, previously it could return a tuple but now it's always returning 1 value. 
I think this code should be deleted:
```
if isinstance(output, tuple):
            return output[0]
        else:
            return output
```

File: models/demos/llama3_70b_galaxy/tt/generator.py
Comment (skhorasganiTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: ok, since i see llama70b-glx's warmup_model_prefill function was already ignoring sampling_params (idk why)

File: models/tt_transformers/tt/generator.py (hunk 17 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py]: From the vLLM side, this function won't be invoked, but from the metal side, it will, and it should check if the model has the sampling module rather that hardcoding it to True. Anyways, in the next PR this call will be removed from here, and moved to a place inside the metal e2e or demo test to call the warmup the same way it is called from vLLM.

File: models/tt_transformers/tt/generator.py
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py]: Will remove it from here, but `simple_vision_demo.py` and gemmas `vision_demo.py` need to be updated to have this check since otherwise tests fail.

**Review by dgomezTT — APPROVED** [review]

**Review by skhorasganiTT — APPROVED** [review]

**Review by arginugaTT — APPROVED** [review]
